python/test/pyJointAxis_TEST.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (C) 2022 Open Source Robotics Foundation
+
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#       http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+from ignition.math import Pose3d, Vector3d
+from sdformat import JointAxis, Error
+import math
+import unittest
+
+
+class JointAxisTEST(unittest.TestCase):
+
+  def test_default_construction(self):
+    axis = JointAxis()
+    self.assertEqual(Vector3d.UNIT_Z, axis.xyz())
+    self.assertFalse(axis.xyz_expressed_in())
+    self.assertAlmostEqual(0.0, axis.damping())
+    self.assertAlmostEqual(0.0, axis.friction())
+    self.assertAlmostEqual(0.0, axis.spring_reference())
+    self.assertAlmostEqual(0.0, axis.spring_stiffness())
+    self.assertAlmostEqual(-1e16, axis.lower())
+    self.assertAlmostEqual(1e16, axis.upper())
+    self.assertAlmostEqual(math.inf, axis.effort())
+    self.assertAlmostEqual(math.inf, axis.max_velocity())
+    self.assertAlmostEqual(1e8, axis.stiffness())
+    self.assertAlmostEqual(1.0, axis.dissipation())
+
+    errors = axis.set_xyz(Vector3d(0, 1, 0))
+    self.assertFalse(errors)
+    self.assertEqual(Vector3d.UNIT_Y, axis.xyz())
+
+    axis.set_xyz_expressed_in("__model__")
+    self.assertEqual("__model__", axis.xyz_expressed_in())
+
+    # expect errors when trying to resolve axis without graph
+    vec3 = Vector3d()
+    self.assertTrue(axis.resolve_xyz(vec3))
```

Comment:
it is counter-intuitive that the resolve function evaluates to `True` on failure (because the Errors vector is not empty) and `False` on success (when Errors is empty).

@azeey is there anything we can do to reverse this? Can we create a custom binding for the `sdf::Errors` vector that reverses this logic?

Reply:
Since `sdf::Errors` is a `std::vector` of `sdf::Error` objects, I think it's automatically converted to python list of `sdf::Error` objects. I'm not sure if it would be possible to reverse the logic for checking if a list is empty. I think it would be easier to change the test assertion:
```python
 def assert_no_errors(errs):
    self.assertFalse(errs)
```
and use `assert_no_errors` here.

Reply:
that helper would certainly make our tests more readable, but I think it would be a defect in our python API if any method that returns an `sdf::Errors` object would resolves to `True` in the presence of errors and `False` when there are no errors. I know that C treats a return code `== 0` as success, but this isn't C. It just seems backwards to me.

~~~
if axis.resolve_xyz(vec3):
    # what do you expect here, success right?
~~~

it seems like a big usability issue to me

Reply:
what about using `len` ?

```python
if len(axis.resolve_xyz(vec3)) == 0:
   ...
```

Reply:
> what about using `len` ?
> 
> ```python
> if len(axis.resolve_xyz(vec3)) == 0:
>    ...
> ```

that's a nice way to do it

maybe I'm overthinking it, but I think now is the time to try to fix this if we think it's worth fixing

Reply:
It does feel backward when written that way because that feels like you would get a `Result` object back from that function call. For someone who expects an `Errors` object, reverting the logic would still lead to confusion because one might do:

```python
errors = axis.resolve_xyz(vec3)
if errors:
  # Handle errors
  print(errors) # or something more important
```

I believe the pythonic way to do error handling is to use exceptions, so a potential solution would be to wrap the function so that it raises an exception if it has an error.

Reply:
> I believe the pythonic way to do error handling is to use exceptions, so a potential solution would be to wrap the function so that it raises an exception if it has an error.

yes, this feels like a more proper way to do things. I think for the time being we can continue using a direct mapping of the API, but I'll add a bullet point to our meta-ticket suggesting error-handling with exceptions

thanks for the great suggestion!

Reply:
Done here https://github.com/ignitionrobotics/sdformat/pull/960/commits/0fae02b8215013d8468c5210312de62b5b7ae531

Reply:
I still see `self.assertTrue(axis.resolve_xyz(vec3))`. Are you planning to 
change it?